[ExpansionPanel] New component

andrzejbk opened the pull request:
This is a pull request for the Expansion Panel Component specified in the [Material UI Design Specifications](https://material.io/guidelines/components/expansion-panels.html). 

There are no tests written for this component so far. However I would like to add them with feedback from the contributors.

There is a demo for two main case of use of the component (standalone Expansion Panel / controlled Expansion Panel - like accordion) included in documentation section. This is apart of the #2863 Issue.

Closes #6303

oliviertassinari commented on docs/src/pages/component-demos/expansion-panels/SimpleExpansionPanel.js:
```diff
@@ -0,0 +1,42 @@
+    width: '100%',
+  },
+  content: {
+    padding: 24,
```
I think that we should be using `theme.spacing.unit` multiple as much as we can. `padding` and `marginTop`.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+}));
+
+const enableTabbing = '0';
+const disableTabbing = '-1';
```
Are those constants really needed 😬 ?

andrzejbk replied:
Just wanted to make the code more descriptive, but you're right! this is not necessary in this case so much 😄

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+  static defaultProps = {
+    disabled: false,
+    expandIcon: null,
+    onChange: () => {},
```
Instead, we use the defensive checks pattern on the rest of the codebase.

andrzejbk replied:
👍

mbrookes replied:
> Instead, we use the defensive checks pattern on the rest of the codebase.

That's a shame. We had started to move away from that in 0.x, as being less verbose in the code, and more descriptive in the docs.

oliviertassinari replied:
The last occurrence of that pattern was removed with https://github.com/callemall/material-ui/commit/1db7c12f23f8202fedf44c3d8eb5e00a92645793. The difference between the two solutions is slight, it's more important to be consistent on the whole code base (minimizing entropy).
Still, this pattern:
 - allows the following logic in userland:
```jsx
<Transition onEnter={open ? this.handleEnter : null} />
```
- is faster, no function allocation, no execution function overhead

mbrookes replied:
I doubt the performance overhead is meaningful, but I agree with your first bullet.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+        className={classes.focusHolder}
+        tabIndex={!disabled && enableTabbing}
+        onKeyUp={e => this.onHeaderKeyUp(e)}
+        onFocus={() => this.handleFocus()}
```
We are creating new function at each render. Not good.

andrzejbk replied:
Great catch! I'm fixing it.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+  isControlled = null;
+
+  onHeaderKeyUp({ keyCode }) {
+    if (keyCode === codes.enter || keyCode === codes.space) {
```
We use the following pattern on the rest of the codebase:
```jss
import keycode from 'keycode';
const code = keycode(event)
if (code === 'enter' || code === 'space') {
```

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+  }
+
+  render() {
+    const { children, classes, className, disabled, unmountOnExit } = this.props;
```
We need to spread the other properties to the root element: Card.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+  /**
+   * Unmounts the child elements after collapse the panel.
+   */
+  unmountOnExit: PropTypes.bool,
```
We miss the default value. `false` sounds good.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+    transition: theme.transitions.create('height', {
+      duration: theme.transitions.duration.shortest,
+    }),
+    padding: [0, theme.spacing.unit * 3],
```
I didn't know that pattern 👍

oliviertassinari commented:
![capture d ecran 2017-08-04 a 01 29 57](https://user-images.githubusercontent.com/3165635/28948195-762ae018-78b4-11e7-9ae6-146a5d6561cd.png)

vs

![capture d ecran 2017-08-04 a 01 30 14](https://user-images.githubusercontent.com/3165635/28948207-817adf40-78b4-11e7-9419-7c5ad688cb9a.png)

andrzejbk commented:
Tkanks you very much for great feedback. I'll take into account all of your suggestions.

mbrookes commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+   */
+  defaultExpanded: PropTypes.bool,
+  /**
+   * If `true`, the panel should be displayed in a disabled state.
```
"...will be displayed in a disabled state"? (Or is there some possibility it won't be?)

andrzejbk replied:
@mbrookes of course "will be" :) And thanks for feedback!

mbrookes commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+   */
+  children: PropTypes.element,
+  /**
+   * Useful to extend the style applied to components.
```
@oliviertassinari I see we're using that wording in other components, but "Useful to" sounds a bit colloquial. Could we find alternate wording, and I'll apply it across v1-beta?

andrzejbk replied:
I agree, so what do you think about "Allows to..." ?

mbrookes replied:
I think we could leave out the prefix ("Useful to", Allows to") altogether as most of the docs comments focus simply on what a prop does. Also, the rest of the 'classes' documentation refers to "overriding" rather than "extending", so perhaps "Override the classes applied to the component's elements" or something?

It might also be helpful to link the word "classes" to the class list (CSS API) further down the page. (component-api/expansion-panel#css-api).

andrzejbk replied:
I made "Allows to", because I have not read your answer before. "Override..." sounds good as well, even though sometims we are "extending" the original class definition by providing only one css property for instance...
Sure, link would be good, is there any simple way to do that now inside the code-comment ?

mbrookes replied:
> is there any simple way to do that now inside the code-comment

Good question! The generated docs are markdown, so I had assumed that a markdown link `[anchor](path)` would just work, but you might have to try it and see.

andrzejbk replied:
@mbrookes, 2 possible solutions:
1) Adding anchor on the word "classes" (I mean `Name` column) requires some changes (I think) in the https://github.com/callemall/material-ui/blob/v1-beta/scripts/generate-docs-markdown.js and creating an specific routines in the `generatePropDescription()` function and adding specific syntax in properity description (similar to `@ignore`) in the code comments (maybe I'm wrong, because I'm not familiar of this part of the code base. Yet ;) ).
2) Simpler way is to add (as you mentioned - thanks) `[title](url)` markup in the property description (in the code), but then the anchor will be placed in Description instead of Name column.
Does the option 2) makes sense for now ?

mbrookes commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,261 @@
+   */
+  onChange: PropTypes.func,
+  /**
+   * Unmounts the child elements after collapse the panel.
```
The description for `bool` props usually begins "If `true`".

andrzejbk commented:
I made additional changes according to the latest requirements.

Additionally I added a new `disableHeaderTypography` prop, in case when the user wants to insert an own header/title component for more advanced cases, like additional summary informations (on header) besides the title.
Please take a look at it and please let me know if this approach makes sense for now. If so, I'll prepare an example for this case of use.

Many thanks for the feedback and many thanks in advance for any further suggestions.

mbrookes commented:
@andrzejbk I commented on the (now outdated) commit: https://github.com/callemall/material-ui/pull/7651#discussion_r131661902

Also, please could you update https://github.com/callemall/material-ui/blob/v1-beta/docs/src/pages/getting-started/supported-components.md ? Thanks.

andrzejbk commented:
@oliviertassinari , @mbrookes, I updated the code according to latest requirements/suggestions.
So if you are able to take a look at it, I could resume working on it.
Many thanks in advance for any further suggestions!

oliviertassinari commented:
@andrzejbk Sorry for the delay. I'm having a look at it now.

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,320 @@
+// @flow
+
```
`// @inheritedComponent Paper`

oliviertassinari replied:
(for the auto doc generation)

oliviertassinari commented on docs/src/pages/component-demos/expansion-panels/SimpleExpansionPanel.js:
```diff
@@ -0,0 +1,52 @@
+          sit amet blandit leo lobortis eget.
+        </Typography>
+      </ExpansionPanel>
+
```
Blank line

oliviertassinari commented on docs/src/pages/component-demos/expansion-panels/SimpleExpansionPanel.js:
```diff
@@ -0,0 +1,52 @@
+
+const styleSheet = theme => ({
+  root: {
+    marginTop: 30,
```
`marginTop: theme.spacing.unit * 3,`

oliviertassinari commented on docs/src/pages/component-demos/expansion-panels/ControlledExpansionPanels.js:
```diff
@@ -0,0 +1,98 @@
+
+const styleSheet = theme => ({
+  root: {
+    marginTop: 30,
```
`marginTop: theme.spacing.unit * 3,`

oliviertassinari commented on src/ExpansionPanel/ExpansionPanel.js:
```diff
@@ -0,0 +1,320 @@
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import keycode from 'keycode';
+import { Paper, Typography, IconButton } from '../';
```
⚠️  tree-shaking. Please import the three of them directly.

oliviertassinari commented:
@andrzejbk Good work so far!

andrzejbk commented:
Many thanks @oliviertassinari for the last review!
Ok, I made a basic fixes and synchronized the branch, however I have a some questions.

Could you please explain me, what details you mean regarding to your ["vs"](https://github.com/callemall/material-ui/pull/7651#issuecomment-320116931) comment ?
Dimensions / corners / shadows / font or maybe missing content in the examples like "Trip name" + "Caribbean cruise" and so on...

ExpansionPanel in this PR has a 804px of width, Google - 725px, besides that, panels heights (collapsed) in Google's examples are not the same. Also borders seems a little bit blured. I'm not sure whether this is an intentional.

Current "vs"
![_r2_c2](https://user-images.githubusercontent.com/25014611/30038904-5de522fa-91ca-11e7-9966-e6d6945f445e.png)

andrzejbk commented:
@oliviertassinari, regarding to markup you proposed - I had a similar feeling to add a separate component which could be called as "<ExpansionHeader />" instead of `headerTitle` prop which are currently included in root `<ExpansionPanel>` component.

The new markup could look like:

# 1
```
<ExpansionPanel>
	<ExpansionHeader> OR <ExpansionSummary>
		...
	</ExpansionHeader>
	<ExpansionContent> OR <ExpansionDetails>
		...
	</ExpansionContent>
	<ExpansionActios>
		...
	</ExpansionActios>
</ExpansionPanel>
```

alternatively (longer "cascaded" names)

# 2
```
<ExpansionPanel>
	<ExpansionPanelHeader> OR <ExpansionPanelSummary>
		...
	</ExpansionHeader>
	<ExpansionPanelContent> OR <ExpansionPanelDetails>
		...
	</ExpansionPanelContent>
	<ExpansionPanelActios>
		...
	</ExpansionPanelActios>
</ExpansionPanel>
```

What do you think?

oliviertassinari commented:
> Could you please explain me, what details you mean regarding to your "vs" comment ?

It looks good this time. My issue was around the overlapping box shadows.

> regarding to markup you proposed

It was more for the underlying dom element than the API but yeah, it's always good to be closer to the DOM elements.

Option 2 looks better but I have some doubt about the implmentation, it will most likely require adding some hacks. Instead, I think that moving the `renderHeader` into his own `ExpansionPanelSummary` component would be great 👍 .

sakulstra commented:
Any updated on this? Can I somehow help here?

andrzejbk commented:
@sakulstra thanks for asking, I'm working on it. I gonna make the next updates within two days. So after that if you are also able to have a look at the code and leave some comments, would be great!

andrzejbk commented:
Latest changes:

- applyed option # 2
- refactor of an `<ExpansionPanel>` component
- moved an ExpansionPanel's (old) header code into new `<ExpansionPanelSummary>` component (can displays the labels and sub labels etc... predefined flex display on the items wrapper)
- added an `<ExpansionPanelDetails>` component (I think, by this the whole expansion panel syntax is more consistent) - can display a content of the ExpansionPanel (we can use it more than once inside the panel)
- added an `<ExpansionPanelActions>` component - can display a button items (just like in Dialog / Card)
- refactored existing examples
- added a new example - "Sub labels and Collumns" (includes a column layout)
- added typings

Left:
- test coverage
- increasing the package size limit ;)

![image](https://user-images.githubusercontent.com/25014611/30742284-b2b3ce68-9f98-11e7-8591-91d1033e7e65.png)

Now the `<ExpansionPanelSummary>` and `<ExpansionPanelDetails>` are initially pre-styled to use a flex display. The only thing what the user nees to customize (If he wants to develop a more than one column) is aplpying an additional style into a column/cell(div) element on the implementation side and set the width rule (See examples). I'm wondering if this will be flexible or not. We can also use a `<Grid>` for this purpose, but not sure which way will be better and shorter.

I'm aware this is not the end of work, so would be great to get the next feedback from you guys.
Thanks in advance for any comments.

oliviertassinari commented:
@andrzejbk I will find some time to have a look at it :)

RainingNight commented:
Is there any progress?

andrzejbk commented:
@oliviertassinari I hope the work on the library goes fine. Did you have a chance to look at the code? Are there any actions that I should take in order to go forward (apart from test coverage) ? Thank you for your support.

oliviertassinari commented:
@andrzejbk I have been focusing on the bugs recently. I haven't found the time to look at the PR. Having a green build will definitely help moving it forward :).

kulakowka commented on docs/src/pages/demos/expansion-panels/ControlledExpansionPanels.js:
```diff
@@ -0,0 +1,102 @@
+  };
+
+  render() {
+    const classes = this.props.classes;
```
You can simplify it:
```jsx
const { classes } = this.props;
```

andrzejbk commented:
A quick status here: thank you all for all feedback and patience. 
I'm gonna move forward this PR, so I'm resuming work on the issing things here. A "green build" should be ready in upcoming week.

andrzejbk commented:
@oliviertassinari I'm done for now. Please let me know if you see any additional things to do. Thanks.

oliviertassinari commented:
@andrzejbk Thank you! I will have a look at the PR this week-end :).

mbrookes commented:
@andrzejbk Thanks for persevering with this! 👍 

Just a few of things: 

1. The original wording of the description is "Expansion panels contain creation flows and allow lightweight editing of an element."; it would be good to copy that. Also for the title "Expansion Panels" (plural).

2. The left border on the "helper text" column in the Sub-labels example seems to have gone wrong.

<img width="817" alt="image" src="https://user-images.githubusercontent.com/357702/32954860-382c3172-cbac-11e7-8355-690c9331a608.png">

vs:

<img width="732" alt="image" src="https://user-images.githubusercontent.com/357702/32955485-04a38876-cbae-11e7-94b9-0f6f0413c190.png">

From the spec it looks like the border is applied to the column rather than the text, at the height of the largest column.

I'm not sure what that question-mark icon in the spec is supposed to represent ("select trip destination" isn't a question), and I don't recall seeing it elsewhere in the spec, but I appreciate you were just copying what's there.

3. The API section needs to list all the sub-components.

4. The ExpansionPanelSummary `onChange` callback needs it's signature documented (`@param`).

Thanks!

d3sandoval commented:
@mbrookes 
> I'm not sure what that question-mark icon in the spec is supposed to represent ("select trip destination" isn't a question)

I think the question mark icon is supposed to represent a "helper text" function. In this kind of UI, a popover might appear to explain more about the "Select trip destination". For the sake of this PR, it can probably just be there for show, and wouldn't be functional in any way.

mbrookes commented:
d3sandoval You might think that, were it not for the fact that the "helper text" is defined in the spec, and shown in that very example:

> Select your destination of choice
> [Learn more]()

(Not the most helpful of helper text, but it _is_ only an example). 

I think the question-mark may be a mistake in the spec, perhaps left from before the designer decide to add the helper text column. I'd be tempted to omit it in favor of the "Learn more" link.

d3sandoval commented:
@mbrookes 🤔 good point! In general, should it be required that all component PRs look exactly like the material design guidelines? Or is the project only looking for spacing, animations and more general consistencies?

mbrookes commented:
@d3sandoval The latter, but most examples in the spec aren't contrary to this, so we borrow heavily from them.

andrzejbk commented:
@mbrookes thanks for the feedback!
1. Fixed

2. a. Fixed - the problem was because I've been using previous notation of declaring multi-values on stylesheet props (`padding: [8, 16]` rather than `padding: '8px 16px'`)

2. b. ...anyway I removed `HelpOutlineIcon` from demo page to prevent potential doubts of why nothing happens on hover on that icon and added fake "Learn more" link (let me know if you have any other suggestions of declaring the `a` element with associated style in here, becouse I'm not sure whether that approach are ok now)

3. Fixed

4. Actually the same `onChange` event (propogated from `Summary`) is exposed in root - ExpansionPanel - element, so I just updated the signature on the root element accordingly, and hid (added @ignore) in ExpansionPanelSummary (same approach as in Tabs/Tab components)

5. Up to date with `v1-beta`

mbrookes commented:
@andrzejbk Thanks, hat looks great! 

Sorry for drip-feeding the feedback, but:

1. Could you perhaps add a description for the second example? Maybe: 

"Multiple columns can be used to structure the content, and a helper text may be added to the panel to assist the user." 

The heading for that example could mention the action buttons too.

2. The examples aren't consistent with each other in the CSS class names applied to the primary and secondary column hedings. The first calls it `title`; the second `header` and `secondary` (while the markdown heading refers to "Sub labels"); and the third example uses `label` and `subLabel`.

The guidelines are inconsistent too. In the [usage](https://material.io/guidelines/components/expansion-panels.html#expansion-panels-usage) section, they refer to "List content" and "Secondary content", but in the [Specs](https://material.io/guidelines/components/expansion-panels.html#expansion-panels-specs) the call it a label.

I think "secondary" is better than "sub" (to me the "optional" caption in the second example is a subheading). "Content" sounds like it's referring to what's in the expanded panel, so I would suggest `heading` and `secondaryHeading` in the CSS, and "Secondary heading" in the markdown, but I'm open to other suggestions. (`label` & `secondaryLabel`?)

3. Please could you add `aria-label`s to the action buttons in the second example. (We aren't consistent in our other examples, some do, some don't, but we hope to improve that.)

@oliviertassinari How should the "Expand" icon button receive its `aria-label`?

Out-of-scope for this PR, but I wondered from what you've seen whether you think ExpansionPanelActions, CardActions and DialogActions have enough in common to merge them into a single Actions component?

oliviertassinari commented:
> ExpansionPanelActions, CardActions and DialogActions have enough in common for us to merge them into a single Actions component?

I have looked into it. I have found room for simplification and implemented them here. But I'm not sure we can share the same components for those three use cases.

> How should the "Expand" icon button receive its aria-label?

I don't know. I haven't taken care of it. I don't know how to handle accessibility with this component.

andrzejbk commented:
Thanks @oliviertassinari for your latest update. It looks perfectly!

Hovewer I have a 2 questions:

1) per last comment from @mbrookes about `aria-labels` on action buttons (pt. 3)  - should we add those labels in case the buttons already have the text content (like Cancel / Save)?... I don't see any similar cases in the code.

2) I'm not sure about the context of changing the Card/Dialog Actions files here, so my question is: should I resolve a current conflicts of those files @oliviertassinari ?

oliviertassinari commented:
@andrzejbk 
1. The component misses the whole accessibility, it's not just about the aria-label. We need to think about it. For your specific use case, I'm wondering if we shouldn't make the icon button invisible from the accessibility tools.
2. I have resolved the conflict this morning. I haven't pushed them yet.

andrzejbk commented:
@oliviertassinari 
1) I'll have a look at it. Is this the last thing blocking this PR ?
2) Thanks!

andrzejbk commented:
Hi,

@oliviertassinari, a small issue encountered after last^ commit :)

**Assumtion**
1) more than **1** column on the Summary & Details
2) column flex-basis: **percentage**

**Current result**
the column alignment of the Details and Summary are inconsistent (see Detailed Expansion Panels example).

I found you've been removing the expand-icon div wrapper which was an absolute - that was an intentional.
Please see the differences:

![obraz](https://user-images.githubusercontent.com/25014611/33037260-37d5cada-ce31-11e7-87a0-8ac162fa4830.png)

Previous code.
```
export const styles = (theme: Object) => {
  ...
  return {
    ...
    action: {
      position: 'absolute',
      top: '50%',
      transform: 'translateY(-50%)',
      right: theme.spacing.unit,
    },
   ...
	'& > :last-child': {   // ... of the container
        paddingRight: theme.spacing.unit * 3,
      },
   ...
}
```
The `:last-child` was the last column of the flex container on the Summary elem with the right-padding added to prevent covering potential text by the expan-button.

The main intention of that approach was to have a both flex containers (Summary and Details) horizontally stretched to the same width.

Now the Summary container width is decreased by the action button space (second image^), so now the potential Summary column size/width may be not the same as the Details have (see the current Detailed Expansion Panel case).

It's mean, when the container has more than one column set by the user as percentage basis, the issue is more significant.

Can we use a previous approach ?
(absolute expand-button + right-pading for the last Summary collumn)

mbrookes commented:
> should we add those labels in case the buttons already have the text content (like Cancel / Save)?

No, your right, my mistake. At one point we used `<div>` as the button container, so the label was needed on all buttons, but that was a long time ago! Now it's only needed for IconButtons and the like where there is no explicit label. Sorry for the confusion.

> The component misses the whole accessibility story

Perhaps not entirely - the keyboard accessibility seems sound, but reading the links that you provided there is far more to it for this component than I realized.

@andrzejbk Sorry for making more work!

mbrookes commented:
> Can we use a previous approach ?

The previous approach works better in terms of assuring that columns line up between Summary and Details, and behaves better at smaller sizes (column headings wrap rather than collapsing under each other), although neither is perfect at mobile widths, but Google doesn't have a good story around that in the spec, so I think that's acceptable for now.

andrzejbk commented:
@mbrookes Thanks for feedback!

Update:
- added basic support (`aria-hidden`, `aria-expanded`) for accesibility tools
- both flex containers (Summary and Details) are horizontally stretched to the same width (see ss)
- the last child column of the Summary now are prestyled - has padding-right == spacing-unit * 4
- invisible icon button invisible from the accessibility tools
- added `containerProps` prop for the `Collapse` element (needed for adding `aria-` props to this component in this case, however I think it wotrh to add this prop regardless of that, earlier or later this prop will be helpful/necassary for other cases)

Current layout - example of edge case:
![image](https://user-images.githubusercontent.com/25014611/33242998-afa2afc8-d2dd-11e7-8bc9-e436efea41be.png)

mbrookes commented:
@andrzejbk Fantastic! I think this is good to merge. @mui-org/core-contributors?

pelotom commented:
Could we add some examples to `components.spec.tsx` to make sure that the TypeScript typings are configured correctly?

andrzejbk commented:
@pelotom thanks, I'm on it.

andrzejbk commented:
@pelotom, done.

pelotom commented:
Great, LGTM 👍🏼

mbrookes commented:
Woohoo! 🚀 

@andrzejbk Thanks so much for all your hard work!

andrzejbk commented:
Thank you all for contribution! It was a great experience!

oliviertassinari commented:
@andrzejbk Awesome job here!

awgeorge commented:
Discovered this: #9396 - any thoughts? Great work on the component though!
